# DESAFIO MÓDULO 3

---

Vamos estudar um dataset do OpenML que contém os dados de 416 pacientes hepáticos e 167 pacientes não hepáticos, coletados no nordeste de Andhra Pradesh. Vale informar que esse dataset contém os dados de 441 homens e 142 mulheres. 

In [1]:
#Primeiro, vamos começar importando todas as bibliotecas necessárias 
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

1. Baixar o arquivo com os dados no link https://www.openml.org/d/1480. O formato do arquivo deve ser CSV.

In [44]:
#Com ele já baixado, vamos ler nosso dataset
df_indian = pd.read_csv(r'C:\Users\Ademir\Desktop\BootCamp IGTI\ML\Módulo 3 - Seleção de Modelos de ML\Desafio\phpOJxGL9.csv')

df_indian.head(10)

V1      V2    V3   V4   V5  V6   V7   V8   V9   V10  Class
0  65  Female   0.7  0.1  187  16   18  6.8  3.3  0.90      1
1  62    Male  10.9  5.5  699  64  100  7.5  3.2  0.74      1
2  62    Male   7.3  4.1  490  60   68  7.0  3.3  0.89      1
3  58    Male   1.0  0.4  182  14   20  6.8  3.4  1.00      1
4  72    Male   3.9  2.0  195  27   59  7.3  2.4  0.40      1
5  46    Male   1.8  0.7  208  19   14  7.6  4.4  1.30      1
6  26  Female   0.9  0.2  154  16   12  7.0  3.5  1.00      1
7  29  Female   0.9  0.3  202  14   11  6.7  3.6  1.10      1
8  17    Male   0.9  0.3  202  22   19  7.4  4.1  1.20      2
9  55    Male   0.7  0.2  290  53   58  6.8  3.4  1.00      1

* V1. Age of the patient. Any patient whose age exceeded 89 is listed as being of age "90".
* V2. Gender of the patient
* V3. Total Bilirubin
* V4. Direct Bilirubin
* V5. Alkphos Alkaline Phosphatase
* V6. Sgpt Alanine Aminotransferase
* V7. Sgot Aspartate Aminotransferase
* V8. Total Proteins
* V9. Albumin
* V10. A/G Ratio Albumin and Globulin Ratio
* Class : hepático: 1 ou ñ hepático: 1

2. Obter informações relativas a número de features e amostras.

In [45]:
#Vamos obter as informações do nosso dataset
df_indian.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 583 entries, 0 to 582
Data columns (total 11 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   V1      583 non-null    int64  
 1   V2      583 non-null    object 
 2   V3      583 non-null    float64
 3   V4      583 non-null    float64
 4   V5      583 non-null    int64  
 5   V6      583 non-null    int64  
 6   V7      583 non-null    int64  
 7   V8      583 non-null    float64
 8   V9      583 non-null    float64
 9   V10     583 non-null    float64
 10  Class   583 non-null    int64  
dtypes: float64(5), int64(5), object(1)
memory usage: 50.2+ KB


3. Verificar a necessidade de tratamento de dados categóricos e valores faltantes.

In [46]:
#É visível que não há dados nulos, mas vamos verificar com os parâmetros
df_indian.describe()

V1          V3          V4           V5           V6  \
count  583.000000  583.000000  583.000000   583.000000   583.000000   
mean    44.746141    3.298799    1.486106   290.576329    80.713551   
std     16.189833    6.209522    2.808498   242.937989   182.620356   
min      4.000000    0.400000    0.100000    63.000000    10.000000   
25%     33.000000    0.800000    0.200000   175.500000    23.000000   
50%     45.000000    1.000000    0.300000   208.000000    35.000000   
75%     58.000000    2.600000    1.300000   298.000000    60.500000   
max     90.000000   75.000000   19.700000  2110.000000  2000.000000   

                V7          V8          V9         V10       Class  
count   583.000000  583.000000  583.000000  583.000000  583.000000  
mean    109.910806    6.483190    3.141852    0.947064    1.286449  
std     288.918529    1.085451    0.795519    0.318492    0.452490  
min      10.000000    2.700000    0.900000    0.300000    1.000000  
25%      25.000000    5.800000    2.600000    0.700000    1.000000  
50%      42.000000    6.600000    3.100000    0.947064    1.000000  
75%      87.000000    7.200000    3.800000    1.100000    2.000000  
max    4929.000000    9.600000    5.500000    2.800000    2.000000

Aparentemente não há nenhum problema.

4. Modelar o SVC e o Random Forest Classifier, com Random Search para sintonia de hiperparâmetros e validação cruzada estratificada, usando as parametrizações abaixo.

5. Parametrização SVC:

    a. Bibliotecas para importação:
    
    b. Kfold estratificado com 10 conjuntos.
    
    c. Métrica de avaliação f1:
    
    d. Parâmetro de kernel:
    
        i. Sigmoidal e RBF.
    e. Parâmetro de regularização C:
    
        ii. Distribuição uniforme variando entre 1 e 10
    f. Random_state = 42.
    
    g. Número de iterações = 5.

In [47]:
#Vamos importar nossas bibliotecas

from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import StratifiedKFold, RandomizedSearchCV
from scipy.stats import randint, uniform
from sklearn.metrics import f1_score, make_scorer

In [52]:
#Agora vamos criar as os parâmetros necessários e hiperparâmetros
classifiers_svm = SVC()
f1 = make_scorer(f1_score)
strat_cv = StratifiedKFold(n_splits=10)

distribuicao = dict(kernel=['sigmoid', 'rbf'],
                   C =uniform(loc = 0, scale= 10))

In [49]:
#Vamos tratar o Male e Female
mapa_gen = {'Male': 1,
           'Female': 0}
df_indian['V2'] = df_indian['V2'].map(mapa_gen)

In [50]:
df_indian.head()

V1  V2    V3   V4   V5  V6   V7   V8   V9   V10  Class
0  65   0   0.7  0.1  187  16   18  6.8  3.3  0.90      1
1  62   1  10.9  5.5  699  64  100  7.5  3.2  0.74      1
2  62   1   7.3  4.1  490  60   68  7.0  3.3  0.89      1
3  58   1   1.0  0.4  182  14   20  6.8  3.4  1.00      1
4  72   1   3.9  2.0  195  27   59  7.3  2.4  0.40      1

In [51]:
#Agora vamos fazer a separação dos nossos dados
labels_indian = np.array(df_indian['Class'])
feature_list = list(df_indian.columns) #estou salvando a disposição das features
data_indian = np.array(df_indian.drop('Class', axis = 1))

In [53]:
#Agora que temos nossos dados, vamos instânciar e treinar o modelo
modelo_random = RandomizedSearchCV(classifiers_svm, distribuicao, cv=strat_cv, scoring=f1, random_state= 42, n_iter=5)
modelo_random.fit(data_indian, labels_indian)

RandomizedSearchCV(cv=StratifiedKFold(n_splits=10, random_state=None, shuffle=False),
                   error_score=nan,
                   estimator=SVC(C=1.0, break_ties=False, cache_size=200,
                                 class_weight=None, coef0=0.0,
                                 decision_function_shape='ovr', degree=3,
                                 gamma='scale', kernel='rbf', max_iter=-1,
                                 probability=False, random_state=None,
                                 shrinking=True, tol=0.001, verbose=False),
                   iid='deprecated', n_iter=5, n_jobs=None,
                   param_distributions={'C': <scipy.stats._distn_infrastructure.rv_frozen object at 0x0000022E3642EC08>,
                                        'kernel': ['sigmoid', 'rbf']},
                   pre_dispatch='2*n_jobs', random_state=42, refit=True,
                   return_train_score=False, scoring=make_scorer(f1_score),
                   verbose=0)

6. Avaliar o resultado da modelagem usando as métricas:

In [54]:
#Vamos analisar o melhor resultado
print('Melhor resultado de f1:', modelo_random.best_score_)
print('\nMelhor configuração de parâmetros:\n', modelo_random.best_params_)
print('\nMelhor estimador:\n', modelo_random.best_estimator_)

Melhor resultado de f1: 0.832818081808181

Melhor configuração de parâmetros:
 {'C': 1.834347898661638, 'kernel': 'rbf'}

Melhor estimador:
 SVC(C=1.834347898661638, break_ties=False, cache_size=200, class_weight=None,
    coef0=0.0, decision_function_shape='ovr', degree=3, gamma='scale',
    kernel='rbf', max_iter=-1, probability=False, random_state=None,
    shrinking=True, tol=0.001, verbose=False)


7. Repetir o processo usando o Random Forest:

    a. Faça a instanciação do Random Forest fixando o random_state = 42:
    
    b. Kfold estratificado com 10 conjuntos.
    
    c. Métrica de avaliação f1:
    
        i. f1 = make_scorer(f1_score)
        
    d. Parâmetro do número de árvores:
    
        i. Distribuição aleatória inteira de valores entre 100 e 200.
        
    e. Parâmetro Bootstrap:
    
        i. Verdadeiro e Falso.
        
    f. Parâmetro Criterion:
    
        i. Gini e Entropy.
        
    g. Random_state = 42.
    
    h. Número de iterações = 5.

In [59]:
#Instânciando nosso problema
classifiers_florest = RandomForestClassifier(random_state = 42)

distribuicao_florest = dict(n_estimators = randint(100, 200),
                           bootstrap = [True, False],
                           criterion = ['gini', 'entropy'])

modelo_random_florest = RandomizedSearchCV(classifiers_florest, distribuicao_florest,
                                          cv = strat_cv, scoring=f1, random_state= 42, n_iter= 5)
modelo_random_florest.fit(data_indian, labels_indian)

RandomizedSearchCV(cv=StratifiedKFold(n_splits=10, random_state=None, shuffle=False),
                   error_score=nan,
                   estimator=RandomForestClassifier(bootstrap=True,
                                                    ccp_alpha=0.0,
                                                    class_weight=None,
                                                    criterion='gini',
                                                    max_depth=None,
                                                    max_features='auto',
                                                    max_leaf_nodes=None,
                                                    max_samples=None,
                                                    min_impurity_decrease=0.0,
                                                    min_impurity_split=None,
                                                    min_samples_leaf=1,
                                                    min_samples_s...
                     

8. Avaliar o resultado da modelagem usando as métricas:

In [60]:
#Vamos analisar o melhor resultado
print('Melhor resultado de f1:', modelo_random_florest.best_score_)
print('\nMelhor configuração de parâmetros:\n', modelo_random_florest.best_params_)
print('\nMelhor estimador:\n', modelo_random_florest.best_estimator_)

Melhor resultado de f1: 0.8019647527623143

Melhor configuração de parâmetros:
 {'bootstrap': True, 'criterion': 'gini', 'n_estimators': 171}

Melhor estimador:
 RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=171,
                       n_jobs=None, oob_score=False, random_state=42, verbose=0,
                       warm_start=False)


## Respondendo o Desafio

1. O objetivo dessa base é avaliar se uma pessoa vai ser um paciente hepático ou não. A feature que determina a classe nessa base é a “V7”.

R: Falso

2. Sobre as informações de número de features e amostras, marque a alternativa CORRETA.

R: A base conta com 583 amostras e 11 features incluindo o atributo da classe.

3. Essa base conta com 3 atributos categóricos. Precisamos fazer o pré processamento desses atributos para que todos sejam numéricos.

R: Falso

4. A base não tem dados faltantes.

R: Verdadeiro

5. Sobre o best_score_ encontrado durante a otimização do SVC, marque a alternativa CORRETA.

R: O valor da f1 encontrado foi aproximadamente 0.83.

6. O best_params_ encontrado durante a otimização do SVC indica que a melhor combinação de parâmetros usa um kernel RBF.

R: Verdadeiro

7. O best_params_ encontrado durante a otimização do SVC indica que C encontrado é aproximadamente 5.

R: Falso

8. A melhor combinação de best_params_ encontrado durante a otimização do SVC é:

R: Kernel RBF com C = 1.834.

9. Sobre o best_score_ encontrado durante a otimização do Random Forest, marque a alternativa CORRETA.

R: O valor da f1 encontrado foi aproximadamente 0.80.

10. O best_params_ encontrado durante a otimização do Random Forest indica que a melhor combinação de parâmetros usa Boostrap True.

R: Verdadeiro

11. O best_params_ encontrado durante a otimização do Random Forest indica que a melhor combinação de parâmetros usa Criterion Entropy.

R: Falso

12. O best_params_ encontrado durante a otimização do Random Forest indica que a melhor combinação de parâmetros conta com quantos estimadores?

R: n_estimators: 171

13. Suponha que nossas duas opções para solução do problema sejam o SVC e o Random Forest que foram modelados anteriormente. Além disso, não é possível rodar o experimento novamente com outros parâmetros. Marque a alternativa que justifica de forma CORRETA a escolha do modelo mais adequado.

R: O SVC é o modelo que deveria ser escolhido já que ele apresenta uma f1 maior.

14. Como os dois modelos apresentaram resultados semelhantes de f1, e considerando apenas essa métrica, podemos entender que a capacidade de generalização dos modelos deve ser semelhante na prática.

R: Verdadeiro

15. Os resultados encontrados do random search para o SVC e o Random Forest representam a melhor configuração possível desses modelos para solução do problema proposto.

R: Falso 